pkg/storage/plugin_adapter.go

```diff
@@ -21,88 +22,61 @@ var _ Store = PluginAdapter{}
 // ResultType on plugin.ResultOptions so that you can just cast the result to
 // the specified type safely.
 type PluginAdapter struct {
-	plugin plugins.StoragePlugin
+	plugin plugins.StorageProtocol
 }
 
 // NewPluginAdapter wraps the specified storage plugin.
-func NewPluginAdapter(plugin plugins.StoragePlugin) PluginAdapter {
-	return PluginAdapter{plugin: plugin}
-}
-
-func (a PluginAdapter) Connect() error {
-	return a.plugin.Connect()
+func NewPluginAdapter(plugin plugins.StorageProtocol) PluginAdapter {
+	return PluginAdapter{
+		plugin: plugin,
+	}
 }
 
 func (a PluginAdapter) Close() error {
-	return a.plugin.Close()
-}
-
-func (a PluginAdapter) Aggregate(collection string, opts AggregateOptions, out interface{}) error {
-	err := a.Connect()
-	if err != nil {
-		return err
+	if closer, ok := a.plugin.(io.Closer); ok {
+		return closer.Close()
 	}
+	return nil
+}
 
-	rawResults, err := a.plugin.Aggregate(opts.ToPluginOptions(collection))
+func (a PluginAdapter) Aggregate(ctx context.Context, collection string, opts AggregateOptions, out interface{}) error {
+	rawResults, err := a.plugin.Aggregate(ctx, opts.ToPluginOptions(collection))
 	if err != nil {
 		return err
 	}
 
 	return a.unmarshalSlice(rawResults, out)
 }
 
-func (a PluginAdapter) EnsureIndex(opts EnsureIndexOptions) error {
-	err := a.Connect()
-	if err != nil {
-		return err
-	}
-
-	return a.plugin.EnsureIndex(opts.ToPluginOptions())
+func (a PluginAdapter) EnsureIndex(ctx context.Context, opts EnsureIndexOptions) error {
+	return a.plugin.EnsureIndex(ctx, opts.ToPluginOptions())
 }
 
-func (a PluginAdapter) Count(collection string, opts CountOptions) (int64, error) {
-	err := a.Connect()
-	if err != nil {
-		return 0, err
-	}
-
-	return a.plugin.Count(opts.ToPluginOptions(collection))
+func (a PluginAdapter) Count(ctx context.Context, collection string, opts CountOptions) (int64, error) {
+	return a.plugin.Count(ctx, opts.ToPluginOptions(collection))
 }
 
-func (a PluginAdapter) Find(collection string, opts FindOptions, out interface{}) error {
-	err := a.Connect()
+func (a PluginAdapter) Find(ctx context.Context, collection string, opts FindOptions, out interface{}) error {
+	rawResults, err := a.plugin.Find(ctx, opts.ToPluginOptions(collection))
```

Comment:
All structs that have Connect are responsible for calling it themselves. It's just the caller who is responsible for calling Close() if defined.